File: packages/flutter/lib/src/face_geometry.dart
import 'types.dart';

/// Frontalness score 0–100. Direct port of the TypeScript getFrontalness().
///
/// Blends three metrics:
/// - Eye span ratio (left-right coverage)
/// - Nose symmetry (how centred the nose is between the eyes)
/// - Vertical tilt (nose-to-eye ratio)
///
/// [kps] must have at least 2 entries: [0]=rightEye [1]=leftEye [2]=noseTip
double getFrontalness(
  List<Keypoint> kps,
  double faceWidth,
  double faceHeight,
) {
  if (kps.length < 2) return 50;

  // ── Horizontal: eye span ──────────────────────────────────────────────────
  final eyeSpan = (kps[1].x - kps[0].x).abs();
  final eyeSpanScore = (eyeSpan / faceWidth * 250).clamp(1.0, 100.0);

  if (kps.length < 3) return eyeSpanScore.roundToDouble();

  // ── Horizontal: nose symmetry ─────────────────────────────────────────────
  final rD = (kps[0].x - kps[2].x).abs();
  final lD = (kps[1].x - kps[2].x).abs();
  final tot = rD + lD;
  final symmetryScore =
      tot > 0 ? (2 * rD.compareTo(lD) < 0 ? rD : lD) / tot * 100 : 50.0;

  // ── Vertical: tilt (up/down) ──────────────────────────────────────────────
  final eyeMidY = (kps[0].y + kps[1].y) / 2;
  final noseRatio = (kps[2].y - eyeMidY) / faceHeight;
  final verticalScore =
      (100 - (noseRatio - 0.35).abs() * 280).clamp(1.0, 100.0);

  return (eyeSpanScore * 0.5 + symmetryScore * 0.3 + verticalScore * 0.2)
      .roundToDouble();
}

/// Non-maximum suppression — remove overlapping detections above [iouThreshold].
List<Detection> nmsFilter(List<Detection> faces, {double iouThreshold = 0.35}) {
  final sorted = [...faces]
    ..sort((a, b) => b.score.compareTo(a.score));
  final kept = <Detection>[];
  for (final face in sorted) {
    if (!kept.any((k) => boxIoU(k.boundingBox, face.boundingBox) > iouThreshold)) {
      kept.add(face);
    }
  }
  return kept;
}

double boxIoU(BoundingBox a, BoundingBox b) {
  final ax2 = a.originX + a.width;
  final ay2 = a.originY + a.height;
  final bx2 = b.originX + b.width;
  final by2 = b.originY + b.height;
  final iw = (ax2 < bx2 ? ax2 : bx2) - (a.originX > b.originX ? a.originX : b.originX);
  final ih = (ay2 < by2 ? ay2 : by2) - (a.originY > b.originY ? a.originY : b.originY);
  if (iw <= 0 || ih <= 0) return 0;
  final inter = iw * ih;
  final union = a.width * a.height + b.width * b.height - inter;
  return union > 0 ? inter / union : 0;
}

File: packages/core/src/face-geometry.ts
/**
 * Frontalness 0–100. Blends three metrics:
 * - Eye span ratio: horizontal eye separation / face width (left-right)
 * - Nose symmetry: how centered the nose is between the eyes (left-right)
 * - Vertical tilt: nose-to-eye vertical distance / face height (up-down)
 *
 * @param kps        Normalised keypoints: [0]=rightEye [1]=leftEye [2]=noseTip
 * @param faceWidth  Bounding box width  normalised to frame (box.width  / frame.width)
 * @param faceHeight Bounding box height normalised to frame (box.height / frame.height)
 */
export function getFrontalness(
  kps: Array<{ x: number; y: number }>,
  faceWidth: number,
  faceHeight: number,
): number {
  if (kps.length < 2) return 50;

  // ── Horizontal: eye span ──────────────────────────────────────────────────
  // Frontal face: eyes ~40% of face width apart → score ~100
  // Profile face: eyes ~5% apart → score ~12
  const eyeSpan = Math.abs(kps[1]!.x - kps[0]!.x);
  const eyeSpanScore = Math.min(100, Math.max(1, (eyeSpan / faceWidth) * 250));

  if (kps.length < 3) return Math.round(eyeSpanScore);

  // ── Horizontal: nose symmetry ─────────────────────────────────────────────
  const rD = Math.abs(kps[0]!.x - kps[2]!.x);
  const lD = Math.abs(kps[1]!.x - kps[2]!.x);
  const tot = rD + lD;
  const symmetryScore =
    tot > 0 ? Math.max(1, (2 * Math.min(rD, lD)) / tot * 100) : 50;

  // ── Vertical: tilt (up/down) ──────────────────────────────────────────────
  // Nose should sit ~35% of face height below the eye midpoint.
  const eyeMidY = (kps[0]!.y + kps[1]!.y) / 2;
  const noseRatio = (kps[2]!.y - eyeMidY) / faceHeight;
  const verticalScore = Math.min(
    100,
    Math.max(1, 100 - Math.abs(noseRatio - 0.35) * 280),
  );

  return Math.round(eyeSpanScore * 0.5 + symmetryScore * 0.3 + verticalScore * 0.2);
}

/** Non-maximum suppression: remove overlapping detections above iouThreshold. */
export function nmsFilter<T extends { boundingBox: BoundingBox; score?: number }>(
  faces: T[],
  iouThreshold = 0.35,
): T[] {
  const sorted = [...faces].sort((a, b) => (b.score ?? 0) - (a.score ?? 0));
  const kept: T[] = [];
  for (const face of sorted) {
    if (!kept.some((k) => boxIoU(k.boundingBox, face.boundingBox) > iouThreshold)) {
      kept.push(face);
    }
  }
  return kept;
}

interface BoundingBox {
  originX: number;
  originY: number;
  width: number;
  height: number;
}

export function boxIoU(a: BoundingBox, b: BoundingBox): number {
  const ax2 = a.originX + a.width;
  const ay2 = a.originY + a.height;
  const bx2 = b.originX + b.width;
  const by2 = b.originY + b.height;
  const iw = Math.max(0, Math.min(ax2, bx2) - Math.max(a.originX, b.originX));
  const ih = Math.max(0, Math.min(ay2, by2) - Math.max(a.originY, b.originY));
  const inter = iw * ih;
  const union = a.width * a.height + b.width * b.height - inter;
  return union > 0 ? inter / union : 0;
}
